### DISCUSSION TITLE
latest/annotators/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#784] [Docs]
# annotators/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/annotators/

**Comment by hafsaJChaudhry** [comment]
how can you use annotators for only certain bounding box id instances?

**Reply by SkalskiP** [reply]
Hi, @hafsaJChaudhry 👋🏻 In principle, it is simple - you must filter out Detections before annotating. But you need to know what you are looking for. 

For example, this will only draw people:

```python
import supervision as sv

image = ...
detections = sv.Detections(...)
detections = detections[detections.class_id == 0]

bounding_box_annotator = sv.BoundingBoxAnnotator()
annotated_frame = bounding_box_annotator.annotate(
    scene=image.copy(),
    detections=detections
)
```

For example, this will draw only an object with `tracker_id = 11`:

```python
import supervision as sv

image = ...
detections = sv.Detections(...)
detections = detections[detections.tracker_id == 11]

bounding_box_annotator = sv.BoundingBoxAnnotator()
annotated_frame = bounding_box_annotator.annotate(
    scene=image.copy(),
    detections=detections
)
```

**Comment by casperdias** [comment]
After try example code of counting-crossing line object, i notice the object BBOX must all through the line so the counter is updated. can it be made that if only the bottom line of bbox passes through then the counter will increase??

**Reply by SkalskiP** [reply]
Hi @casperdias 👋🏻 Yup, it is possible! You need to adjust the value of `triggering_anchors`. Initiate your `LineZone` like this.

```python

line_zone = sv.LineZone(
    start=<START_POINT_VALUE>,
    end=<END_POINT_VALUE>, 
    triggering_anchors=[sv.Position.BOTTOM_LEFT, sv.Position.BOTTOM_RIGHT]
)
```

This way, `LineZone` will only trigger if the bottom edge crosses the line. `LineZone` does not have the docs yet... but you can look at the code responsible for this behavior [here](https://github.com/roboflow/supervision/blob/4a58c0b72633ee9ca3cb496b2cdd55a5ae47dbec/supervision/detection/line_counter.py#L34).

**Reply by casperdias** [reply]
TYSM, I tried it and it worked. about the counter, is there any way to access the result of the LineZone.Trigger like true or false?

**Reply by SkalskiP** [reply]
You mean if you can access information if specific detections cross line in this specific video frame?

**Reply by casperdias** [reply]
yes I already found it so no problem

**Reply by SkalskiP** [reply]
That's awesome to hear!

**Comment by wcoolers** [comment]
Hi, is there a way I can customize the labels, e.g, put names of my choice on objects detected? Also, is there a way to implement object re-identification, in case an object goes out of frame and comes back in? Thanks a lot

**Reply by SkalskiP** [reply]
> is there a way to implement object re-identification

It is on our long-term TODO list but, it is currently not possible with supervision.

> is there a way I can customize the labels

Absolutely. All you need to do is pass the `labels` list as the argument to the `LabelAnnotator.annotate` method. The one important thing is that `len(labels)` must be equal to `len(detections)`. 

```python
labels = [
    <YOUR_LABEL>
    for _
    in detections
]

label_annotator = sv.LabelAnnotator(text_position=sv.Position.CENTER)
annotated_frame = label_annotator.annotate(
    scene=image.copy(),
    detections=detections,
    labels=labels
)
```

**Comment by casperdias** [comment]
Hi, is this library support protocol like RTSP or HTTP Live streaming to detect object?

**Reply by SkalskiP** [reply]
Supervision on its own does not yet support detection on live stream, but you can combine supervision with [inference](https://github.com/roboflow/inference) to do it.

**Comment by millzee01** [comment]
Hi, is there a way to copy the detected mask to a transparent image?  

i.e create a transparent image and copy the mask object form the original image to the new transparent image thus making a new image contain just the masked detections?

**Reply by SkalskiP** [reply]
Hi @millzee01. Yes. Do you want to save each mask as separate image? Or put all together on the same image?

**Reply by SkalskiP** [reply]
Or do you prefer a result like this?

![download - 2024-03-21T171219 625](https://github.com/roboflow/supervision/assets/26109316/752de216-1e6e-453e-a3ed-ad41cd07e7e6)

**Comment by tobieabel** [comment]
Hi - I've been using the HeatMap annotator and added in a simple homograph function so that xy coordinates from a typical CCTV camera view can be translated into a top down view for annotating onto a floor plan, as this is the most typical use-case I've found for heat mapping in industry.  Works pretty well, and you can use your polygonZone webapp to generate the homography mappings.  Is this something you would consider putting into supervision main branch?

**Reply by SkalskiP** [reply]
Hi @tobieabel 👋🏻 Would you be interested in preparing a PR with a POC of your solution?

**Reply by tobieabel** [reply]
@SkalskiP Sure, I'm a self taught amateur when it comes to coding and not done a PR before so you might need to guide me through it, but happy to submit the change I made, it only took a few lines of openCv code

**Reply by SkalskiP** [reply]
To start with, follow [contributor guidelines](https://github.com/roboflow/supervision/blob/develop/CONTRIBUTING.md). Using those instructions, you should be able to submit the first PR than I'll try to provide you with comments and further instructions.

**Reply by tobieabel** [reply]
@SkalskiP ok, I'll go through that and submit the PR this week and see what you think.  Its just a small change and actually very similar to your transformation code on your blog on speed estimation on the 19th Jan this year

**Reply by SkalskiP** [reply]
@tobieabel awesome! I can't wait to take a look at the PR.

**Comment by aberg2323** [comment]
TraceAnnotator Question: how do i set the trace length as something other than the default. I am having trouble implementing this change in my code.

**Reply by LinasKo** [reply]
Hi @aberg2323 👋 

I presume you've tried setting the `trace_length` argument already. What happens if you set it to a higher or lower value?

**Reply by aberg2323** [reply]
Unfortunately, I do not see a difference when i change the trace_length argument to a different value. Any other thoughts?

**Reply by LinasKo** [reply]
Gotcha. I'll verify tomorrow and get back to you.

**Reply by aberg2323** [reply]
Thank you so much for your help! I appreciate it greatly.

**Reply by LinasKo** [reply]
Hi @aberg2323,

I can't reproduce the issue. Here's the code I used - if there's an issue on your end, it may help debug it. And if the issue persists, let me know.

```python
import supervision as sv
from supervision.assets import VideoAssets, download_assets
from ultralytics import YOLO


download_assets(VideoAssets.VEHICLES)

model = YOLO("yolov8s")

video_info = sv.VideoInfo.from_video_path(VideoAssets.VEHICLES.value)
frame_gen = sv.get_video_frames_generator(VideoAssets.VEHICLES.value)
tracker = sv.ByteTrack()
trace_annotator_1 = sv.TraceAnnotator(trace_length=45, color=sv.Color.BLUE)
trace_annotator_2 = sv.TraceAnnotator(trace_length=30, color=sv.Color.GREEN)
trace_annotator_3 = sv.TraceAnnotator(trace_length=15, color=sv.Color.WHITE)
trace_annotator_4 = sv.TraceAnnotator(trace_length=7, color=sv.Color.YELLOW)
trace_annotator_5 = sv.TraceAnnotator(trace_length=3, color=sv.Color.RED)


with sv.VideoSink("out.mp4", video_info) as sink:
    for frame in frame_gen:
        result = model(frame)[0]
        detections = sv.Detections.from_ultralytics(result)
        detections = tracker.update_with_detections(detections)
        annotated_frame = frame.copy()
        trace_annotator_1.annotate(annotated_frame, detections)
        trace_annotator_2.annotate(annotated_frame, detections)
        trace_annotator_3.annotate(annotated_frame, detections)
        trace_annotator_4.annotate(annotated_frame, detections)
        trace_annotator_5.annotate(annotated_frame, detections)
        sink.write_frame(annotated_frame)

```

**Reply by aberg2323** [reply]
Thank you so much! I was able to change the length with that code.

**Reply by LinasKo** [reply]
Best of luck in your project! 😉

**Comment by gusmendess** [comment]
I want to annotate my true positives as green bounding boxes, false positives as yellow bounding boxes and my false negatives as red bounding boxes for better understanding of my model. Of course comparing with my true bounding boxes annotations. How can i do that?

**Reply by SkalskiP** [reply]
Hi @gusmendess, do you mean something similar to this: https://x.com/skalskip92/status/1795194852121969104?

**Reply by gusmendess** [reply]
@SkalskiP Yes but in my case i want to compare my model predictions with my true labels. I want to my bounding boxes that are true positives to be color green, bboxes that are false positives color yellow and bboxes that are supposed to be detected but weren't as red bboxes
